Move join to its own pkg

Jaanki opened the pull request:
... and decouple cobra stuff

Signed-off-By: Janki Chhatbar <jchhatba@redhat.com>

Depends on https://github.com/submariner-io/submariner-operator/pull/1724

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Jaanki committed:
Move join to its own pkg

... and decouple cobra stuff

Signed-off-By: Janki Chhatbar <jchhatba@redhat.com>

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,554 @@
+
+var status = cli.NewStatus()
+
+func SubmarinerCluster(subctlData *datafile.SubctlData, jo Options) error{
```
Let's use the new structure introduced by https://github.com/submariner-io/submariner-operator/pull/1717 once it's merged.

```suggestion
func SubmarinerCluster(brokerInfo broker.Info, options Options) error{
```

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,554 @@
+		if len(answers.ColorCodes) > 0 {
+			jo.ColorCodes = answers.ColorCodes
+		}
+	}
```
L90-132 are related to the CLI and should be in _cmd/subctl/join.go_. We shouldn't do any prompting here.

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,554 @@
+		}
+	}
+
+	clientConfig, err := restConfigProducer.ClientConfig().ClientConfig()
```
The `restConfigProducer` should be passed in.

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,554 @@
+	}
+
+	clientConfig, err := restConfigProducer.ClientConfig().ClientConfig()
+	utils.ExitOnError("Error connecting to the target cluster", err)
```
We should return errors and let the caller decide to exit.

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,554 @@
+limitations under the License.
+*/
+
+package join
```
Perhaps we should put this in the `deploy` package and name the entry function `Operator`. So similar to `deploy.Broker`, callers would reference `deploy.Operator`.

Jaanki replied:
`deploy.Operator` is another function to deploy operator. May be this can be `deploy.Submariner`?

tpantelis replied:
yeah `deploy.Submariner` sounds reasonable. Its responsibility should basically be to deploy the `Submariner` and `ServiceDiscovery` CRs. I think the gateway node labeling should be separate - we can put that code as part of _cmd/subctl/join.go_ at least for now and perhaps move it to its own lib package under _pkg_ later.

Jaanki commented:
This is still in WIP and not ready for review at all.A lot of work is still pending.

tpantelis commented on internal/exit/exit.go:
```diff
@@ -33,10 +33,12 @@ func OnError(err error) {
-	fmt.Fprintln(os.Stderr, message)
-	printVersion()
-	os.Exit(1)
+func WithMessage(message string, err error) {
```
I think we should leave this function as is - there are cases where we want to just print a message and exit in the absence of an error. For the case where we want to augment an error with an additional message, I suggest an `OnErrorWithMessage` function.

tpantelis commented:
https://github.com/submariner-io/submariner-operator/pull/1724 is adding more code to **join** so we'll need to take that into account here.

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+	PreRunE: restConfigProducer.CheckVersionMismatch,
+	Run: func(cmd *cobra.Command, args []string) {
+		err := checkArgumentPassed(args)
+		exit.OnError(err)
```
We should print the error message. I suggest we change `checkArgumentPassed` to:

```
func checkArgumentPassed(args []string) {
	if len(args) == 0 {
                 exit.WithMessage("The broker-info.subm file argument generated by 'subctl deploy-broker' is missing")
	}
}
```

and

```suggestion
		checkArgumentPassed(args)
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+
+		subctlData, err := datafile.NewFromFile(args[0])
+		exit.WithMessage("Error loading the broker information from the given file", err)
+		fmt.Printf("* %s says broker is at: %s\n", args[0], subctlData.BrokerURL)
```
This should use the new `broker.ReadInfoFromFile`. Also let's use the CLI status here:

 ```suggestion
                 status := cli.NewReporter()
                 
                 brokerInfo, err := broker.ReadInfoFromFile(args[0])
                 exit.OnError(status.Error(err, "Error loading the broker information from file %q", args[0]))

                 status.Success("Loaded the broker information from file %q", args[0])
                 status.Success("The broker is located at %q", brokerInfo.BrokerURL)
```

Jaanki replied:
Commit https://github.com/submariner-io/submariner-operator/pull/1726/commits/6fc60d5e759dbebb4f84a4806616706bfaa5de6a for new `broker.ReadInfoFromFile` and https://github.com/submariner-io/submariner-operator/pull/1726/commits/e05535f023bd0dd66743675e987261a0b36c17a3 for `reporter.Interface`

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+
+		if joinFlags.ClusterID == "" {
+			joinFlags.ClusterID, err = askForClusterID()
+			exit.WithMessage("Error collecting information", err)
```
Let's report to the CLI status. Also in other places below.

```suggestion
                         exit.OnError(status.Error(err, "Error collecting cluster ID"))
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+
+		clientConfig, err := restConfigProducer.ClientConfig().ClientConfig()
+		_, clientset, err := restconfig.Clients(clientConfig)
+		exit.WithMessage("unable to set the Kubernetes cluster connection up", err)
```
We can use `ForCluster` here and get the `Producer` from that:

```suggestion
                 config, err := restConfigProducer.ForCluster()
		exit.OnError(status.Error(err, "Error creating REST config"))

		clientProducer, err := client.NewProducerFromRestConfig(config)
		exit.OnError(status.Error(err, "Error creating client producer"))
```

Jaanki replied:
I think I am already using `Producer` in the commit https://github.com/submariner-io/submariner-operator/pull/1726/commits/51e1409953a40a1a88af883e71406fec38dfba59

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+		_, clientset, err := restconfig.Clients(clientConfig)
+		exit.WithMessage("unable to set the Kubernetes cluster connection up", err)
+
+		gatewayNodesPresent, err := join.ListGatewayNodes(clientConfig)
```
`ListGatewayNodes` is only used here so move to this file. Also pass the client interface from the `Producer`.

```suggestion
		gatewayNodesPresent, err := join.ListGatewayNodes(clientProducer.ForKubernetes())
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+		gatewayNodesPresent, err := join.ListGatewayNodes(clientConfig)
+		if err != nil {
+			exit.WithMessage("Error getting gateway node", err)
+		}
```
```suggestion
		exit.OnError(status.Error(err, "Error retrieving the gateway nodes"))
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+		var gatewayNode struct{Node string}
+		// If not Gateway nodes present, get all worker nodes and ask user to select one of them as gateway node
+		if !gatewayNodesPresent {
+			allWorkerNodeNames, err := join.GetAllWorkerNodeNames(clientset)
```
```suggestion
			allWorkerNodeNames, err := join.GetAllWorkerNodeNames(clientProducer.ForKubernetes())
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+		// If not Gateway nodes present, get all worker nodes and ask user to select one of them as gateway node
+		if !gatewayNodesPresent {
+			allWorkerNodeNames, err := join.GetAllWorkerNodeNames(clientset)
+			if err != nil {
+				exit.WithMessage("error listing worker nodes", err)
+			}
+			gatewayNode, err = askForGatewayNode(allWorkerNodeNames)
+			exit.WithMessage("Error getting gateway node", err)
```
```suggestion
			exit.OnError(status.Error(err, "Error retrieving the worker nodes"))

			gatewayNode, err = askForGatewayNode(allWorkerNodeNames)
			exit.OnError(status.Error(err, "Error collecting the gateway node"))
```

skitt replied:
This can seem surprising but it is idiomatic Go now... (Basically, `status.Error()` and `exit.OnError()` do nothing if there is no error.)

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+			exit.WithMessage("Error getting gateway node", err)
+		}
+
+		brokerInfo, err := broker.ReadInfoFromFile(broker.InfoFileName)
```
This is now retrieved above.

```suggestion
```

Jaanki replied:
Yes, its a left over,

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+		brokerInfo, err := broker.ReadInfoFromFile(broker.InfoFileName)
+
+		err = join.SubmarinerCluster(*brokerInfo, joinFlags, restConfigProducer, cli.NewReporter(), gatewayNode)
+		exit.WithMessage("Error joining cluster", err)
```
We should assume the status is complete and the returned error has been reported already.

```suggestion
		exit.OnError(err)
```

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+			}
+
+			_, err := join.IsValidClusterID(str)
+			fmt.Printf("error is %s", err)
```
We don't need to print here.

```suggestion
```

Jaanki replied:
yes, It was a leftover of a local debug effort.

tpantelis commented on cmd/subctl/join.go:
```diff
@@ -0,0 +1,238 @@
+	autodetect := false
+	prompt := &survey.Confirm{
+		Message: fmt.Sprintf("Do you want to autodetect %s CIDR?", name)}
+
```
This is a change in behavior from the current `join` command where it tried to discover the network details and prompt the user if needed. Since the intention of this PR is to restructure the code, I think we should keep the same behavior. So that means the network discovery should be initiated here and not in the library component.

Jaanki replied:
I think the library should autodetect the networks if no information is passed. So this is what is implemented here. For third party user, it can't be prompted to enter network details if autodetection fails. And hence these changes. The design is
"if network details are not provided via subctl flags, prompt the user. if the details are provided but doesnot match the autodetected, use the provided details." Library always expects the some or empty (strings) details to be provided.

tpantelis replied:
I'm not sure I follow. The current behavior is:
 
If the service CIDR is provided by the user, use it (but print a warning if it doesn't match the discovered CIDR). Otherwise, if the CIDR was discovered, use it (although it actually leaves the Submariner field empty so the operator will detect it). Otherwise prompt the user for the CIDR. 

This changes the behavior to first ask the user if they want to discover/autodetect the CIDR if one is not provided. If not then it prompts. Otherwise if it's not autodetected, it fails. Regardless of whether or not we want this change, it is out of scope of this PR.

tpantelis replied:
Proposed changes with https://github.com/submariner-io/submariner-operator/pull/1726/commits/ce48e99552685255569ecf3c7a7e84ea4b2a5e63.

tpantelis commented on internal/cluster/cluster.go:
```diff
@@ -0,0 +1,53 @@
+	if clusterName != nil {
+		return *clusterName, nil
+	}
+	return "", nil
```
This function is merely wrapper for `restConfigProducer.ClusterNameFromContext` to handle a a nil return value. But there really isn't any value in returning a pointer from `ClusterNameFromContext`. Usually an empty string is used to indicate  not found which this wrapper function does. So I think we should change `ClusterNameFromContext` to return a non-pointer `string` and thus eliminate the need for this wrapper function.

Jaanki replied:
That would mean changes in other unrelated file. We can address this as part of separate PR once this is merged.

tpantelis replied:
OK then I think this function really belongs in `restconfig.Producer`, `GetClusterID()`.

tpantelis replied:
Proposed changes with https://github.com/submariner-io/submariner-operator/pull/1726/commits/f53f3f25addc1f9140028c0d91e4bbc0f01f97e0.

tpantelis commented on internal/cluster/cluster.go:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	return true, nil
+}
```
Add newline.

tpantelis commented on internal/cluster/networks.go:
```diff
@@ -0,0 +1,68 @@
+		networkDetails.Show()
+	}
+
+	return networkDetails, nil
```
Thus function is merely a wrapper for `network.Discover` with the possibly unwanted side-effect of printing the details. Thus it doesn't seem generally reusable so I think we should keep it in the cmd front-end.

Jaanki replied:
I want to keep cmd front-end as simple and subctl specific as possible. It should just be cobra stuff.. All other projects try to follow it.

tpantelis replied:
But it isn't generally reusable. As I said it has the hidden,  possibly unwanted side-effect of printing the details. This should be done by the front-end.

tpantelis replied:
Moved as part of https://github.com/submariner-io/submariner-operator/pull/1726/commits/ce48e99552685255569ecf3c7a7e84ea4b2a5e63.

tpantelis commented on internal/cluster/networks.go:
```diff
@@ -0,0 +1,68 @@
+		return nd.PodCIDRs[0], true, nil
+	} else {
+		return "", false, fmt.Errorf("could not determine cluster network")
+	}
```
This function doesn't seem generally reusable since it determines a result based on input which seems tied to a  specific case, especially the warning message. So, again, I think thus belongs in the cmd front-end.

Jaanki replied:
Ditto

tpantelis replied:
Same argument - it has side effects and thus is not generally reusable. It is specific to the behavior of the join command.

tpantelis replied:
Moved as part of https://github.com/submariner-io/submariner-operator/pull/1726/commits/ce48e99552685255569ecf3c7a7e84ea4b2a5e63.

tpantelis commented on internal/cluster/networks.go:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+func GetServiceCIDR(serviceCIDR string, nd *network.ClusterNetwork, status reporter.Interface) (cidrType string, autodetected bool, err error) {
```
Same as above.

Jaanki replied:
Ditto

tpantelis replied:
Moved as part of https://github.com/submariner-io/submariner-operator/pull/1726/commits/ce48e99552685255569ecf3c7a7e84ea4b2a5e63.

tpantelis commented on internal/constants/constants.go:
```diff
@@ -20,6 +20,8 @@ package constants
 	SubmarinerBrokerAdminSA = "submariner-k8s-broker-admin"
+	SubmarinerNamespace    = "submariner-operator"
+	SubmarinerGatewayLabel = "submariner.io/gateway"
+	TrueLabel = "true"
```
Minor nit: There's other places where we use the literal "true" so we should be consistent and use the constant everywhere. However I don't really think it's necessary to make a constant for "true".

Jaanki replied:
I only moved the constant from some other place to here. It reads out nice in the whole line context. About changing it to other places, its out of the scope of this PR. Trying to keep the PR join specific.

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+// labels the specified worker node as a gateway node
+func HandleLabels(clientset kubernetes.Interface, gatewayNode struct{Node string}) error {
```
It seems odd to pass an anonymous struct - why not just pass a string?

```suggestion
func LabelGateways(clientset kubernetes.Interface, gatewayNode struct{Node string}) error {
```

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,147 @@
+func HandleLabels(clientset kubernetes.Interface, gatewayNode struct{Node string}) error {
+
+	if gatewayNode.Node == "" {
+		fmt.Printf("no gateway nodes specified, selecting one of the worker node as gateway node")
```
Please remove.

Jaanki replied:
Remove what?

tpantelis replied:
The `fmt.Printf`.  These reusable lib functions should not have the side effect of printing info.

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,147 @@
+	if err != nil {
+		return errors.Wrap(err, "Error labeling the gateway node")
+	}
+	return nil
```
```suggestion
	return errors.Wrap(err, "error labeling node %q as a gateway", gatewayNode.Node)
```

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// GetAllWorkerNodeNames returns all worker nodes
+func GetAllWorkerNodeNames(clientset kubernetes.Interface) ([]string, error) {
```
"Node" is already part of the package name.

```suggestion
func GetWorkerNames(clientset kubernetes.Interface) ([]string, error) {
```

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// ListGatewayNodes returns all nodes labelled as gateway
+func ListGateways(config *rest.Config) (bool, error) {
```
This should take a `kubernetes.Interface` instead of `rest.Config`.

However it's advertised to list and return all nodes labeled as gateway but it actually returns whether or not there are any gateways. I think it should do as advertised. Also it shouldn't print out the gateways as a side-effect.

tpantelis commented on pkg/deploy/submariner.go:
```diff
@@ -0,0 +1,213 @@
+	return nil
+}
+
+func ServiceDiscovery(clientProducer client.Producer, jo WithJoinOptions, brokerInfo broker.Info, brokerSecret *v1.Secret, imageOverrides map[string]string, status reporter.Interface) error {
```
Please move this to a separate _service_discovery.go_ file. 

Also I think it would be cleaner to define a separate `ServiceDiscoveryOptions`.

tpantelis commented on pkg/deploy/submariner.go:
```diff
@@ -0,0 +1,213 @@
+	"strings"
+)
+
+func Submariner(clientProducer client.Producer, jo WithJoinOptions, brokerInfo broker.Info, brokerSecret *v1.Secret, netconfig globalnet.Config, imageOverrides map[string]string, status reporter.Interface) error {
```
I think we should define a specific `SubmarinerOptions` instead of using the broader `WithJoinOptions`.

Jaanki replied:
THere are few options common to both submariner and servicediscovery

tpantelis replied:
Yes but  most aren't and some don't apply to either `Submariner` or `ServiceDiscovery` but to the broker join process. I think it would be cleaner to define separate `Options` types for each and to avoid unnecessary coupling.

tpantelis commented on pkg/discovery/globalnet/globalnet.go:
```diff
@@ -460,3 +462,41 @@ func ValidateExistingGlobalNetworks(kubeClient kubernetes.Interface, namespace s
+	})
+
+	return retryErr // nolint:wrapcheck // No need to wrap here
+}
```
newline

tpantelis commented on pkg/discovery/globalnet/globalnet.go:
```diff
@@ -460,3 +462,41 @@ func ValidateExistingGlobalNetworks(kubeClient kubernetes.Interface, namespace s
+
+func AllocateAndUpdateGlobalCIDRConfigMap(clusterID string, brokerAdminClientset *kubernetes.Clientset, brokerNamespace string,
+	netconfig *Config, status reporter.Interface) error {
+	status.Start("Discovering multi cluster details")
```
Since we leave it up to the caller to end the status, I think we let the caller start it.

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,216 @@
+	"strings"
+)
+
+func SubmarinerCluster(brokerInfo broker.Info, jo deploy.WithJoinOptions, restConfigProducer restconfig.Producer, status reporter.Interface, gatewayNode struct{Node string}) error {
```
Since we join a broker...  

Also pass a `client.Producer`.

Also I think we should define an `Options` struct separate from `deploy`.

```suggestion
func Broker(brokerInfo broker.Info, options Options, clientProducer client.Producer, status reporter.Interface, gatewayNode struct{Node string}) error {
```

Jaanki replied:
> Since we join a broker...
> 
We join a submariner cluster to a broker cluster.
> Also pass a `client.Producer`.
> 
`restConfigProducer` is needed by `DetermineID` so passing it here and getting `clientProducer` at L66 below.
 
> Also I think we should define an `Options` struct separate from `deploy`.
Where?

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,216 @@
+		}
+	}
+
+	if brokerInfo.IsConnectivityEnabled() && jo.LabelGateway {
+		err := nodes.HandleLabels(clientProducer.ForKubernetes(), gatewayNode)
+		if err != nil {
+			return status.Error(err, "Unable to set the gateway node up")
+		}
+	}
```
I think labeling gateways should be separate from joining a broker.

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,139 @@
+	"time"
+)
+
+// labels the specified worker node as a gateway node
+func LabelGateways(clientset kubernetes.Interface, gatewayNode struct{Node string}) error {
```
It actually labels one gateway, Also pass a string instead of an anonymous struct.

```suggestion
// LabelGateway the specified node as a gateway or if the node is not specified, find a worker node to label.
func LabelGateway(clientset kubernetes.Interface, nodeName string}) error {
```

tpantelis commented on internal/nodes/nodes.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// ListGateways returns all nodes labelled as gateway
+func ListGateways(clientset kubernetes.Interface) (bool, error) {
```
Retrying this comment as it got removed by the last push...

This should take a `kubernetes.Interface` instead of `rest.Config`.

However it's advertised to list and return all nodes labeled as gateway but it actually returns whether or not there are any gateways. I think it should do as advertised. Also it shouldn't print out the gateways as a side-effect.

tpantelis committed:
Move cluster.DetermineID to restconfig.Producer.GetClusterID

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Validate the cluster ID at the cmd level

...instead of in join.SubmarinerCluster since it's obtained from user input.
This also preserves the current behavior where we prompt the user for the
cluster ID if the CLI argument is invalid.

Subsequently we can then pass in a client.Producer rather than a
restconfig.Producer to join.SubmarinerCluster improving reusability.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Preserve check requirements behavior

...where it always checks and reports failed requirements but
only fails the join if the ignoreRequirements flag is false.

Also log the failures to the status reporter instead of fmt.Printf.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Preserve CIDR detection behavior

The new implementation changed the behavior to first ask the user if they
want to discover/autodetect the CIDR if it wasn't provided. If the user
chooses no then it prompts for the CIDR. Otherwise if it's not
autodetected, it fails.

Changed it to preserve the existing behavior in the cmd front-end where
if the CIDR is provided by the user, use it. Otherwise, if the CIDR
wasn't autodetected, prompt the user for the CIDR.

The library function, join.SubmarinerCluster, simply passes the CIDRs as
is to the Submariner CR and assumes the caller has determined them
appropriately or left one or more empty to let the operator controller
discover it.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Modify ListGateways to return the node names

...as advertised instead of whether or not there are any gateways.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Remove gateway labeling from join.SubmarinerCluster

The join cmd front-end already handles most of the work by
retrieving the worker nodes, prompting the user etc so it may
as well do the last part (nodes.LabelAsGateway) instead of passing
params to join.SubmarinerCluster to do it. A third-party client
could also easily handle the labeling via the functions provided
by the nodes pkg.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Define SubmarinerOptions type in lieu of using WithJoinOptions

...with deploy.Submariner as only a subset of WithJoinOptions fields apply.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

Jaanki commented on cmd/subctl/join.go:
```diff
@@ -69,6 +71,10 @@ var joinCmd = &cobra.Command{
 		clientProducer, err := client.NewProducerFromRestConfig(clientConfig)
 		exit.OnError(status.Error(err, "Error creating the client producer"))
 
+		networkDetails := getNetworkDetails(clientProducer, status)
```
I find 2 issues with this:
1. I strongly believe that `cmd/subctl` should just strictly be `cobra` stuff and nothing else.
2. we are expecting `pkg/join` to know `*CIRD` before hand. So `join` is not smart enough to figure out these details. Any third part code using `join`  would also have to pre-determine these values and can't reply on `join`.

skitt replied:
Where should `survey`-based code reside? The library itself can’t rely on that, since it can’t expect to have any way to interact with an end-user.

tpantelis replied:
> I find 2 issues with this:
> 
> 1. I strongly believe that `cmd/subctl` should just strictly be `cobra` stuff and nothing else.

By `cobra` stuff I assume you mean processing arguments and prompting the user.  That's what this code is doing. We call `network.Discover` here in order to determine if we should prompt the user for any CIDRs if one is not provided.

> 2. we are expecting `pkg/join` to know `*CIRD` before hand. So `join` is not smart enough to figure out these details. Any third part code using `join`  would also have to pre-determine these values and can't reply on `join`.

They can by simply calling `network.Discover`. But they don't need to do this. If the CIDRs aren't provided then the operator controller will determine them via `network.Discover`.

Jaanki replied:
`survey-` code should reside here. It should ask for CIDR here. `join` should `network.Discover`, get the CIDR and warn users if it is not same as user-specified adn update the `ServiceCIDRAutoDetected` accordingly. But I see you removed that field completely. So I guess, it won't make sense to just warn the users if the provided and detected values won't match.

tpantelis replied:
> `survey-` code should reside here. It should ask for CIDR here. `join` should `network.Discover`, get the CIDR and warn users if it is not same as user-specified 

That's what we're doing here. It would be wasteful and unnecessary to perform `network.Discover` twice during `subctl join`. We don't currently have any other users of this API so I think it's reasonable at this point to assume the caller has done their due diligence wrt the CIDRs, if they provide them. We can always revisit based on feedback if we ever do have another user. Also the operator does this same check and logs a warning anyway.

>adn update the `ServiceCIDRAutoDetected` accordingly. But I see you removed that field completely. So I guess, it won't make sense to just warn the users if the provided and detected values won't match.

`ServiceCIDRAutoDetected` was used to pass an empty CIDR to the operator if it was auto-detected and was really unnecessary.

Jaanki commented on cmd/subctl/join.go:
```diff
@@ -69,6 +71,10 @@ var joinCmd = &cobra.Command{
 		exit.OnError(status.Error(err, "Error creating the client producer"))
 
+		networkDetails := getNetworkDetails(clientProducer, status)
+		determinePodCIDR(networkDetails, status)
```
L58-66 needs to be deleted.

tpantelis replied:
yeah - forgot that. Thanks.

Jaanki commented on internal/nodes/nodes.go:
```diff
@@ -70,34 +71,28 @@ func GetAllWorkerNames(clientset kubernetes.Interface) ([]string, error) {
+	return toNodeNames(workerNodes), nil
+}
+
+func toNodeNames(nodes *corev1.NodeList) []string {
```
Rename to `getNodeName`

Jaanki commented on pkg/deploy/join.go:
```diff
@@ -23,7 +23,6 @@ type WithJoinOptions struct {
 	IpsecDebug                    bool
 	SubmarinerDebug               bool
 	OperatorDebug                 bool
-	LabelGateway                  bool
```
A flag can't just be removed. It needs to be deprecated first and then removed in the next release.

tpantelis replied:
This struct is introduced by this PR so it doesn't matter.

Jaanki commented on pkg/deploy/join.go:
```diff
@@ -17,16 +17,16 @@ package deploy
 	PreferredServer               bool
 	ForceUDPEncaps                bool
-	NatTraversal                  bool
+	NATTraversal                  bool
```
Again, not sure if renaming flags is ideal to do without any warning.

skitt replied:
This isn’t renaming a flag, is it? The flag names are determined by the flag setup code (`PersistentFlags().StringVar()` etc.).

tpantelis replied:
Same

tpantelis replied:
yeah this is just the struct field which is new.

tpantelis committed:
Define ServiceDiscoveryOptions type in lieu of using WithJoinOptions

...with deploy.ServiceDiscovery as only a subset of WithJoinOptions
fields apply.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Refactor deploy.WithJoinOptions to join.Options

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented on pkg/join/join.go:
```diff
@@ -0,0 +1,216 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+func SubmarinerCluster(brokerInfo *broker.Info, options *Options, clientProducer client.Producer,
```
Wondering if we should rename this function. The _Submariner_ part seems redundant since we already know we're Submariner 😄 Some possibilities depending on perspective:

1. `join.Cluster`
2. `join.ToBroker`
3. `join.ToClusterSet`

I'm leaning towards 2. @Jaanki @skitt WDYT?

Jaanki replied:
I am leaning towards 1. It joins a cluster (to a broker or to a clusterSet) or lets go overboard with a 4th option `join.ClusterToBroker`.

tpantelis replied:
It depends on one's perspective and preference. `join.ClusterToBroker` would be the most complete. @skitt do you guys have any strong preference?

skitt replied:
Remembering Java, I like 1, because it names the result we’re after; the caller shouldn’t care that joining a cluster means joining a broker. However since Go doesn’t have polymorphism, and we’re explicitly passing in a `broker.Info`, `ClusterToBroker` makes sense to me — if we end up having other variants which don’t use a broker, we’ll say `ClusterToSomethingElse`.

tpantelis replied:
yeah makes sense. We'll go with `ClusterToBroker`. I think with just `Cluster`, it could be interpreted as "_joining to a cluster_" instead of "_joining a cluster to something_".

tpantelis committed:
Rename SubmarinerCluster to ClusterToBroker

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

dfarrell07 committed:
Merge branch 'devel' into move_join

skitt committed:
Merge branch 'devel' into move_join

Jaanki committed:
Merge branch 'devel' into move_join
